azkaban-exec-server/src/test/java/azkaban/dag/DagBuilderTest.java

```diff
@@ -79,6 +79,27 @@ private NodeBuilder createNodeBuilder(final String name) {
     return this.dagBuilder.createNode(name, mock(NodeProcessor.class));
   }
 
+  @Test
+  public void build_should_throw_exception_when_circular_dependency_is_detected() {
+    // given
+    final NodeBuilder nodeBuilder1 = createNodeBuilder("nb1");
+    final NodeBuilder nodeBuilder2 = createNodeBuilder("nb2");
+    final NodeBuilder nodeBuilder3 = createNodeBuilder("nb3");
+    nodeBuilder2.addParents(nodeBuilder1);
+    nodeBuilder3.addParents(nodeBuilder2);
+    nodeBuilder1.addParents(nodeBuilder3);
+
+    // when
+    final Throwable thrown = catchThrowable(() -> {
+      this.dagBuilder.build();
+    });
+
+    System.out.println("Expect exception: " + thrown);
```

Comment:
should we use System.err ?

Reply:
No. This is for debugging purposes. It's not an error.

Reply:
IMHO there's no use leaving test debug output like this in place. If one needs similar thing at some point it's easy to add. Getting rid of extra output keeps the test logs cleaner. Just something for you to consider..

Reply:
Gradle writes such output to a log file instead of the console by default. I agree that writing large verbose debug outputs can make searching for failures in the log more difficult. However, most of the test failures can be reproduced locally and it is easy to see which tests fail from console or IDE. 

I left this log in the code to make it easier to see the exception message when running in an IDE.

On balance, I think it is OK to leave some reasonable amount of debugging information in the test output log after considering the tradeoffs.